rust/src/api.rs

```diff
@@ -145,13 +146,61 @@ pub fn decode_invoice(invoice: String) -> Result<LightningInvoice> {
 pub enum Event {
     Ready,
     Offer(Option<Offer>),
+    WalletInfo(Option<WalletInfo>),
+}
+
+#[derive(Clone)]
+pub struct WalletInfo {
+    pub balance: Balance,
+    pub bitcoin_history: Vec<BitcoinTxHistoryItem>,
+    pub lightning_history: Vec<LightningTransaction>,
+}
+
+impl WalletInfo {
+    async fn build_wallet_info() -> Result<WalletInfo> {
+        let balance = wallet::get_balance()?;
+        let bitcoin_history = WalletInfo::get_bitcoin_tx_history().await?;
+        let lightning_history = wallet::get_lightning_history().await?;
+        Ok(WalletInfo {
+            balance,
+            bitcoin_history,
+            lightning_history,
+        })
+    }
+
+    async fn get_bitcoin_tx_history() -> Result<Vec<BitcoinTxHistoryItem>> {
+        let tx_history = wallet::get_bitcoin_tx_history()
+            .await?
+            .into_iter()
+            .map(|tx| {
+                let (is_confirmed, timestamp) = match tx.confirmation_time {
+                    None => (false, OffsetDateTime::now_utc().unix_timestamp() as u64),
+                    Some(blocktime) => (true, blocktime.timestamp),
+                };
+
+                BitcoinTxHistoryItem {
+                    sent: tx.sent,
+                    received: tx.received,
+                    txid: tx.txid.to_string(),
+                    fee: tx.fee.unwrap_or(0),
+                    is_confirmed,
+                    timestamp,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        Ok(tx_history)
+    }
 }
 
 #[tokio::main(flavor = "current_thread")]
 pub async fn run_ldk(stream: StreamSink<Event>) -> Result<()> {
     tracing::debug!("Starting ldk node");
     let background_processor = wallet::run_ldk().await?;
     stream.add(Event::Offer(offer::get_offer().await.ok()));
+    stream.add(Event::WalletInfo(
+        WalletInfo::build_wallet_info().await.ok(),
```

Comment:
🤔 Actually, why do we have special one-off cases here? Shouldn't it be sufficient to schedule the tasks below?

Reply:
> 🤔 It's weird to swallow the error. Maybe we should log internally for these and return `Option`s instead.

You are right, I liked this approach because the code looks much cleaner, however we will loose the error messages. I will try to rework.

> 🤔 Actually, why do we have special one-off cases here? Shouldn't it be sufficient to schedule the tasks below?

I thought it would be nice to have the wallet info synced at least once before marking the app ready.